src/lib/optimizer/strategy/predicate_placement_rule.cpp

```diff
@@ -343,53 +368,27 @@ void PredicatePlacementRule::_push_down_traversal(const std::shared_ptr<Abstract
         return;
       }
 
-      if (diamond_origin_node->input_count() == 1) {
```

Comment:
What's the reasoning for removing this branch? I wonder if it was dead code or in which scenario a diamond origin node would have only a single input count.

Reply:
This was not dead code, but it gets replaced with shorter and more elegant code down below (line 377ff), which involves a temporary root node.

Previously, we could not simply recurse with `_push_down_traversal` at the diamond's origin node because some previously collected predicates might no longer be evaluable at the diamond's origin node. Since `_push_down_traversal` could not place such predicates appropriately, we had to manually take care of it. By inserting a  temporary root node, however, we can rely on `_push_down_traversal`.